Use GSON objects to build JSON and not string operations

emlaver opened the pull request:
*What*
Building JSON using string operations is causing several defects. Use the GSON library to assemble the object instead.
Issue #101 

*How*
- Replace string building implementation with GSON objects in Database, Search, and CouchDatabaseBase class

*Testing*
New test cases to assert that invalid JSON will throw a bad request.

reviewer @ricellis
reviewer @rhyshort

ricellis commented on src/main/java/com/cloudant/client/api/Database.java:
```diff
@@ -1029,135 +1029,67 @@ public void ensureFullCommit() {
-        }
+        JsonObject indexObject = new JsonObject();
+        //Add selector json object
+        indexObject.addProperty("selector", selectorJson);
```
Does this work? I though the supplied selector JSON was of the form `"selector" : {}`

emlaver replied:
This is the toString value for `indexObject` after adding the selector object:
`{"selector":"{\"Person_name\":\"Alec Guinness\",\"Movie_year\":{\"$gt\":1960}}"}`

ricellis replied:
Isn't the index object going to end up with: `"selector" : "\"selector\" : {...}` because the string`selectorJson` at this point includes `"selector" : ` at the beginning.
If you run the cloudantServiceTest you'll see what I mean.

emlaver replied:
I see your point but will add that the `"selector" :` at the beginning does not always exist in the `selectorJson` string.  This is a problem either way and will need to be handled.

ricellis commented on src/main/java/com/cloudant/client/api/Search.java:
```diff
@@ -369,7 +369,12 @@ public Search counts(String[] countsfields) {
+        drillDownArray.add(fieldNamePrimitive);
+        JsonPrimitive fieldValuePrimitive = new JsonPrimitive(fieldValue);
+        drillDownArray.add(fieldValuePrimitive);
+        uriBuilder.query("drilldown", drillDownArray);
```
`drilldown` can be specified multiple times, looks like the previous code was incorrect in this regard.
It should be `uriBuilder.query("drilldown", drillDownArray, false)` to not replace the parameter. I'm guessing this will be impacted by #167 though. Maybe it is worth moving this out into another issue as it is a bit unrelated.

emlaver replied:
Added as issue #174

emlaver commented:
@ricellis I've added logic to try and deserialize the JSON in `selectorJson`.   If it's valid, the selector object will be added to the final JSON object.  If it's not valid, check if the string starts with `"selector"`.  If it does start with selector, then create a substring and deserialize.  If not, still try to deserialize and throw an exception if it's invalid json syntax.

Also, after changing the original string implementation, I didn't add the index options field and sort into the final JSON object.  Those are also added in 66d039b.

ricellis commented on src/main/java/com/cloudant/client/api/Database.java:
```diff
@@ -1029,135 +1029,116 @@ public void ensureFullCommit() {
+                        JsonPrimitive.class);
+                sortArray.add(json);
+                json = client.getGson().fromJson(sort.getOrder().toString(),
+                        JsonPrimitive.class);*/
```
Remove commented out code

ricellis commented on src/main/java/com/cloudant/client/api/Database.java:
```diff
@@ -1029,135 +1029,116 @@ public void ensureFullCommit() {
+                sortArray.add(sortObject);
             }
-            so.append("]");
+            //sortObject.add("sort", sortArray);
```
More commented out code

ricellis commented on src/main/java/com/cloudant/client/api/Database.java:
```diff
@@ -1029,135 +1029,116 @@ public void ensureFullCommit() {
-            finalbody.append(",")
-                    .append("\"use_index\": ")
-                    .append(options.getUseIndex());
+            JsonArray useIndexArray = new JsonArray();
```
This `useIndexArray` is never used.

ricellis commented on src/main/java/com/cloudant/client/api/Database.java:
```diff
@@ -1029,135 +1029,116 @@ public void ensureFullCommit() {
+            JsonArray useIndexArray = new JsonArray();
+            JsonPrimitive element = new JsonPrimitive(options.getUseIndex());
+            useIndexArray.add(element);
+            indexObject.add("use_index", getGson().fromJson(options.getUseIndex(), JsonElement.class));
```
I think this is the only line needed, but it might be worth changing `JsonElement.class` to `JsonArray.class` since we should probably break if the `getUseIndex()` String doesn't turn into an array.

I think this also highlights a place in `FindByIndexOptions` that is still doing String JSON operations. That is creating the `useIndex. Probably worth changing the internal type for `useIndex` in `FindByIndexOptions` into a `JsonArray` and `return useIndex.toString();` from the `getUseIndex()` method. Although the `JsonArray` -> `String` -> `JsonArray` is slightly cumbersome it avoids exposing the internal JSON choice on the public `getUseIndex()` method. An alternative might be to use a reduced visibility method that returns the `JsonArray` directly to avoid the journey via String.

emlaver replied:
If I change `JsonElement.class` to `JsonArray.class`, I get an error on test cases when the useIndex is not in array form.
I think it's a better idea to see how I can change the internal type for `useIndex` into a `JsonArray`.

emlaver replied:
Looks like I would need to use a `public` access modifier to call an internal `getGsonUseIndex()` method in `Search`.   How about just changing the return type for `FindByIndexOptions.getUseIndex()` to `JsonArray`? I know it's very possible that it would break code, but we could just add a comment that `toString` can be used?

rhyshort commented on src/main/java/com/cloudant/client/api/model/FindByIndexOptions.java:
```diff
@@ -142,7 +148,13 @@ public Integer getReadQuorum() {
 
+    /**
+     * For use in {@link Database#getFindByIndexBody(String, FindByIndexOptions)}
+     */
```
Missing `@return` from java doc

emlaver replied:
I've removed this method in aba42c9.

ricellis commented on src/main/java/com/cloudant/client/api/model/FindByIndexOptions.java:
```diff
@@ -142,7 +148,13 @@ public Integer getReadQuorum() {
+    /**
+     * For use in {@link Database#getFindByIndexBody(String, FindByIndexOptions)}
+     */
+    public JsonArray getUseIndexAsJsonArray() {
```
I'm reluctant to add a public method to this API class.
Maybe just make the field public?

emlaver replied:
After chatting over Slack, I decided on calling GSON's fromJson method to parse the JSON.  Fixed in aba42c9

ricellis commented:
Can you also add an entry to `CHANGES.md` for this?

ricellis commented:
+1

rhyshort commented:
+1
